ENH: Vectorising np.linalg.qr

czgdp1807 opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Fixes #11741 

**Comments**

This is my first PR to numpy. I thought of pushing the early changes so that reviews can be provided as soon as possible and I can keep updating my local branch frequently, keeping the things up to date.

As of now I have just initialized the framework for implementing QR decomposition at C level. I haven't made any change to Python code, will do that once the C code looks good. See, https://github.com/numpy/numpy/issues/11741#issuecomment-851927795

czgdp1807 commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3233,6 +3233,28 @@ static void
+    n = (fortran_int)dimensions[1];
+    excess = m - n;
+
+    if (init_@lapack_func@(&params, m, n, nrhs)) {
+        LINEARIZE_DATA_t a_in, q_out, r_out;
+
+    }
```
As of now this is empty. I will this in the upcoming commits. `a_in` is the square matrix (should we allow non-square matrices as well? QR decomposition is defined only for square matrix). `q_out` is the Q component of the decomposition and `r_out` is the R component of the decomposition.

czgdp1807 commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3379,6 +3402,13 @@ static char lstsq_types[] = {
     NPY_CDOUBLE, NPY_CDOUBLE, NPY_DOUBLE, NPY_CDOUBLE, NPY_DOUBLE, NPY_INT, NPY_DOUBLE,
 };
 
+static char qr_types[] = {
+    NPY_FLOAT,   NPY_FLOAT,   NPY_FLOAT,
+    NPY_DOUBLE,  NPY_DOUBLE, NPY_DOUBLE,
+    NPY_CFLOAT,  NPY_FLOAT,  NPY_FLOAT,
+    NPY_CDOUBLE, NPY_DOUBLE, NPY_DOUBLE,
+};
```
The first entry is for the type of input matrix and the second and third are for Q and R.

czgdp1807 commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3587,6 +3617,15 @@ GUFUNC_DESCRIPTOR_t gufunc_descriptors [] = {
+    },
+    {
+        "qr",
+        "(m,m),->(m,k),(k,m)",
```
`(m, m)` - input dimension, `(m, k)` for Q and `(k, m)` for R. `k = min(m, n)`.

eric-wieser replied:
I don't think you're allowed to have dimensions that appear only on the rhs.

Also, shouldn't this be `(m, n)` not `(m, m)`?

eric-wieser replied:
Unfortunately, there are four variants of `qr`:

https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/linalg/linalg.py#L788-L791

You'll need a definition for each one.

mattip replied:
I think you need two definitions each for reduced and raw. For instance, reduced will have two functions: reduced_n with `(m,n) -> (m,n),(n,n)` and reduced_m  `(m,n)->(m,m),(m,n)` (you also have excess commas?) - see lstsq_m and lstsq_n for examples. Then the python code chooses between them 

https://github.com/numpy/numpy/blob/6790873334b143117f4e8d1f515def8c7fdeb9fb/numpy/linalg/linalg.py#L2296-L2300

mattip replied:
.. assuming you allow non-square matrices

czgdp1807 replied:
> Also, shouldn't this be (m, n) not (m, m)?

Yeah. Actually, I thought that QR decomposition is defined for square matrices so I kept `(m, m)`. However, since, `numpy` allows non-square matrices as well, so we should change the input dimension to `(m, n)`.

> You'll need a definition for each one.

Ah, I see. Thanks for the hint. I will make the changes.

czgdp1807 commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3587,6 +3617,15 @@ GUFUNC_DESCRIPTOR_t gufunc_descriptors [] = {
+        "(m,m),->(m,k),(k,m)",
+        "QR decomposition of last two dimensions and broadcast to the rest.\n"\
+        "The input matrix must be square.\n",
+        4, 1, 2,
```
1 input and 2 output matrices.

czgdp1807 commented:
Please let me know if the above additions require any modifications. A bunch of questions,

1. How to access the `lapack_routine` (used in Python definition of QR (np.linalg.qr)) for doing the decomposition in C code?
2. How to decide which types would be filled in arrays ending with `_types` suffix (such as, `lstsq_types`, `svd_1_3_types`, etc.)?

ping @eric-wieser @mattip 

Thank you. Apologies if I missed out on any guidelines.

eric-wieser commented:
Note you'll need `qr` versions of all of these `lstsq` functions:

https://github.com/numpy/numpy/blob/a436fb99bc1f430693e2212045ed1f1a1fd0dd4e/numpy/linalg/umath_linalg.c.src#L2838-L3234

Not just the final definition.

czgdp1807 commented:
> Note you'll need qr versions of all of these lstsq functions:

Sure. So, can we say that this is the kind of pattern to be followed for every new function added to `umath_linalg.c.src`?

mattip commented:
> ... this is the kind of pattern to be followed for every new function added ...

It depends on the underlying LAPACK/BLAS functionality. In general, I think it would be better to add functionality to SciPy and not to Numpy. In fact, you may be able to reuse some ideas or code from the [scipy.linalg.qr](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.qr.html) implementation.

czgdp1807 commented:
In the recent push I have written a generic implementation of QR decomposition. I will add specific ones for 'r', 'raw', 'economic' and 'reduced_m' and 'reduced_n' separately in future commits.

czgdp1807 committed:
ENH: r, raw, economic now accept stacked matrices

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,6 +3252,289 @@ static void
 /**end repeat**/
 
+/* -------------------------------------------------------------------------- */
+                 /* qr (modes - r, raw, economic) */
```
I don't think you should mention `economic` here as it's just a deprecated spelling of `raw` with some weirdness that C doesn't care about.

czgdp1807 replied:
Okay. I will remove it. The C implementation will remain the same as `r`, `raw` and `economic` use the computation done by this function.

czgdp1807 committed:
WIP

czgdp1807 committed:
ready for formal addition of tests and documentation update

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3535,6 +3551,528 @@ static void
+        BEGIN_OUTER_LOOP_3
+            int not_ok;
+            linearize_@TYPE@_matrix(params.A, args[0], &a_in);
+            size_t safe_min_m_n = fortran_int_min(m, n);
+            size_t safe_m = m;
```
You should be able to compute this outside the loop

czgdp1807 replied:
Sure. Will make the change. Thanks for pointing out.

czgdp1807 committed:
moving intialization outside the loop

czgdp1807 committed:
existing tests passed

czgdp1807 committed:
tests for stacked inputs added

czgdp1807 committed:
documentation updated

czgdp1807 committed:
linting issue fixed

czgdp1807 committed:
removed trailing white spaces

czgdp1807 commented on numpy/linalg/linalg.py:
```diff
@@ -2177,7 +2174,7 @@ def lstsq(a, b, rcond="warn"):
     If `a` is square and of full rank, then `x` (but for round-off error)
     is the "exact" solution of the equation. Else, `x` minimizes the
-    Euclidean 2-norm :math:`||b - ax||`. If there are multiple minimizing 
+    Euclidean 2-norm :math:`||b - ax||`. If there are multiple minimizing
```
```suggestion
    Euclidean 2-norm :math:`||b - ax||`. If there are multiple minimizing 
```

czgdp1807 committed:
resolved linting issues

czgdp1807 committed:
resolved linting issues

czgdp1807 committed:
documentation fix

czgdp1807 commented:
Can someone please restart [this build](https://travis-ci.com/github/numpy/numpy/jobs/511280895) on Travis CI? It seems like the build error-ed due to some issue on Travis servers. Thanks.

czgdp1807 commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,160 +3269,1009 @@ static void
+                   fortran_int m,
+                   fortran_int n)
+{
+    printf("Inside zungqr\n");
```
Will remove it. Noticed it in the CI outputs. Apologies.

czgdp1807 replied:
Meanwhile, please let me know if any updates are required in this PR. I will bundle the changes together and push here.

mattip commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1710,6 +1712,65 @@ def test_mode_all_but_economic(self):
+                self.check_qr_stacked(B)
+                self.check_qr_stacked(A + 1.j*B)
+
+        np.random.set_state(curr_state)
```
I don't think we reset the random number state after tests, that is on each test to seed as it sees fit.

czgdp1807 replied:
Are the seeds globally set for testing purposes? Or in other words can I assume that the tests above would be static in nature?

czgdp1807 replied:
I will remove this setting and resetting of seeds. I noticed a call `np.random.seed(1234)` at the top of the file. So, I believe the seeds are set globally for tests in this file. Thanks for pointing out.

mattip replied:
Each test that cares is expected to seed the random number generator it uses. I am not sure what you mean by "static in nature". We could at some point consider creating a random number generator fixture, but it seems to me better for any test that cares to call `np.random.seed(xxx)`, which is more verbose but also much clearer.
```

czgdp1807 replied:
Thanks for a detailed overview.

> I am not sure what you mean by "static in nature"

By static tests I meant that are the testing data same for every run of the tests. Specifically, if the seed for each test (globally or locally) is not set then testing may vary everytime we run the tests. However, if the seeds are fixed then the testing won't vary. I observed that `np.random.seed` is called at least once inside the above file and hence, the tests should be static. I was resetting the seeds after my tests because I thought that if there are any seed sensitive tests inside the above file then they shouldn't change. However, I realized that it isn't required, so removed that part.

mattip commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,160 +3269,1009 @@ static void
+    size_t safe_n = n;
+
+    size_t a_size = safe_m * safe_n * sizeof(@ftyp@);
+    size_t tau_size = safe_min_m_n * sizeof(@ftyp@);
```
Can these overflow if m, n are too large?

czgdp1807 replied:
If `a_size` will overflow then possibly the error would also be there in other functions defined in this file, since memory is allocated using similar logic everywhere. In other words, this is a broader issue and can arise in any function if a very large matrix is passed. Please let me know if I am missing out on something here. Thanks.

czgdp1807 replied:
For example, in the below init_lapack function for SVD, a_size is calculated in the same manner,

https://github.com/numpy/numpy/blob/21838b4b21d8c702730aba80131fe6ee5fe5116f/numpy/linalg/umath_linalg.c.src#L2479

I think if `tau_size` will overflow then `a_size` would definitely overflow.

czgdp1807 replied:
I am resolving this for now. Please let me know if a separate issue should be opened for this as it covers many functions in this file.

czgdp1807 committed:
removed debug prints and addressed review

czgdp1807 committed:
Added release notes entry

czgdp1807 commented on doc/release/upcoming_changes/19151.improvement.rst:
```diff
@@ -0,0 +1,6 @@
+`np.linalg.qr` accepts stacked matrices as inputs
+-------------------------------------------------
+
+`np.linalg.qr` is able to produce results for stacked matrices as inputs.
+Moreover, the implementation of QR decomposition has been shifted to C
+from Python.
```
Please let me know if this entry needs any improvements. Thanks.

eric-wieser commented on doc/release/upcoming_changes/19151.improvement.rst:
```diff
@@ -0,0 +1,6 @@
+`np.linalg.qr` accepts stacked matrices as inputs
```
```suggestion
`numpy.linalg.qr` accepts stacked matrices as inputs
```
sphinx doesn't know how to link ``np.``

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -904,83 +921,64 @@ def qr(a, mode='reduced'):
+    signature = 'DD->D' if isComplexType(t) else 'dd->d'
+    extobj = get_linalg_error_extobj(_raise_linalgerror_qr_r_raw)
+    q = gufunc(a, tau, signature=signature, extobj=extobj)
+    r = triu(a[..., :mc, :])
```
Can you add a comment explaining `mc`?

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,160 +3269,1008 @@ static void
+    NPY_CDOUBLE, NPY_CDOUBLE, NPY_DOUBLE, NPY_CDOUBLE
+};
+
+/*  A,           b,           rcond,      x,           resid,      rank,    s,        */
```
can you add a comment like this for `qr` too?

czgdp1807 replied:
Sure. Thanks for the suggestion.

czgdp1807 committed:
Addressed reviews

czgdp1807 committed:
fixed title line in release notes

czgdp1807 commented:
The tests are passing. Please let me know if any further updates are needed from my side on this PR. Thanks.

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,160 +3269,1005 @@ static void
+
+    }
+
+    params->LWORK = fortran_int_max(fortran_int_max(1, n),
+                                    work_count > 0 ? work_count : -work_count);
```
What does negative work_count mean here>

czgdp1807 replied:
I have implemented [this line](https://github.com/numpy/numpy/pull/19151/files#diff-a5aa16fdab19a25c2c3c370e09bf02278a3d3118f5596724bffd91ec51419f65L973) in C.

czgdp1807 replied:
I will try removing this to see if things work without it. If so, then I will push the changes here.

czgdp1807 replied:
I think we may need this. Tests are failing on 32bit windows after this change. Let me push once again after re-introducing this part.

czgdp1807 replied:
The failed test on windows is `TestMatmul.test_vector_matrix_values` which I believe is probably unrelated to my changes.

mattip replied:
Maybe another instance of numpy/numpy#18500?

czgdp1807 replied:
Tests on Azure passed now. I am resolving this for now, since, the C implementation is now inline with python implementation. Let me know if there are any other concerns regarding this part.

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,160 +3269,1005 @@ static void
 
 /**end repeat**/
 
-#pragma GCC diagnostic pop
```
I'm a little alarmed that github can't produce a correct diff here. I'll check out your branch and see if it reproduces locally. If it does, then maybe we should just put `qr` further up in the file to prevent this mess.

czgdp1807 replied:
Please let me know if the diff is not good on your system (locally). I will shift qr above along with other changes. Thanks.

czgdp1807 replied:
I shifted qr above in this file. The diff seems better.

czgdp1807 replied:
I am resolving this for now. Please restart the discussions if needed. Thanks.

czgdp1807 committed:
Take work_count as it is

czgdp1807 committed:
shifted qr code in umath_linalg.c.src above

czgdp1807 committed:
updated method for finding optimal work_count

czgdp1807 commented:
ping @eric-wieser Please let me know if any updates are required to this PR. Thanks.

czgdp1807 commented:
@eric-wieser @mattip Are there any improvements to be made here? Please let me know.

mattip commented:
@eric-wieser any more review here?

czgdp1807 committed:
factored out common init parts for sorgqr, dorgqr

czgdp1807 committed:
refactoring complete

czgdp1807 commented:
I have factored out common code in initialization for `sorgqr`, `dorgqr` and their complex counterparts. Let me know if I can do more to reduce the size of the PR. Thanks.

pearu commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1710,6 +1712,61 @@ def test_mode_all_but_economic(self):
+        dts = [np.float32, np.float64, np.complex64]
+        for size in sizes:
+            for dt in dts:
+                a1, a2, a3, a4 = [normal(size=size), normal(size=size),
+                                  normal(size=size), normal(size=size)]
+                b1, b2, b3, b4 = [normal(size=size), normal(size=size),
+                                  normal(size=size), normal(size=size)]
+                A = np.asarray([[a1, a2], [a3, a4]], dtype=dt)
+                B = np.asarray([[b1, b2], [b3, b4]], dtype=dt)
```
Use:
```
A = np.random.normal(size=(2, 2) + size).astype(dt)
B = ..
```

pearu commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1710,6 +1712,61 @@ def test_mode_all_but_economic(self):
+                A = np.asarray([[a1, a2], [a3, a4]], dtype=dt)
+                B = np.asarray([[b1, b2], [b3, b4]], dtype=dt)
+                self.check_qr_stacked(A)
+                self.check_qr_stacked(B)
```
This check is superfluous, remove.

pearu commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1710,6 +1712,61 @@ def test_mode_all_but_economic(self):
+        normal = np.random.normal
+        sizes = [(3, 4), (4, 3), (4, 4), (3, 0), (0, 3)]
+        dts = [np.float32, np.float64, np.complex64]
+        for size in sizes:
```
I would suggest introducing another loop over outer sizes to improve coverage of batched testing. For instance:
```
for outer_size in [(2, 2), (2,), (2, 3, 4)]:
    for size in sizes:
         for dt in dts:
             A = np.random.normal(size=outer_size + size).astype(dt)
             ...
```

eric-wieser replied:
It would be better to handle these loops via `pytest.parametrize`

czgdp1807 replied:
> It would be better to handle these loops via `pytest.parametrize`

Yes. I will do this wherever possible.

pearu commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1710,6 +1712,61 @@ def test_mode_all_but_economic(self):
+
+        normal = np.random.normal
+        sizes = [(3, 4), (4, 3), (4, 4), (3, 0), (0, 3)]
+        dts = [np.float32, np.float64, np.complex64]
```
Suggest include `np.complex128` to dts list.

eric-wieser replied:
Or written another way:
```suggestion
        dts = [np.single, np.double, np.csingle, np.cdouble]
```

pearu reviewed (changes requested):
This review part covers only the Python part of the PR: mostly nits, clarifications, and request to improve testing coverage.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -904,83 +921,68 @@ def qr(a, mode='reduced'):
-        raise LinAlgError('%s returns %d' % (routine_name, results['info']))
+        gufunc = _umath_linalg.qr_r_raw_n
+
+    signature = 'D->D' if isComplexType(t) else 'd->d'
```
No, this is not a gufunc signature, this is a type signature. The terminology is confusing, but this is consistent with the rest of the file.

For reasons lost to time, all of the linalg functions promote single precision to double precision. However, the underlying gufuncs do support single precision.

czgdp1807 replied:
I found out that we can do `F->f` for single precision floating point numbers. The problem is that, `_commonType` always sets `t = double` and hence, the input array always gets converted to `double` type (no matter whether the original input is `float` or `double` type). Let me know, if I should include separate signatures for `float` type arrays. The max we will be required to do is to keep `t` and `result_t` inside the definition of `np.linalg.qr`.

Though I feel that this is a more broader problem and might not be specific to this function only. May be a separate PR can be made to solve such type problems in all those functions under `np.linalg` where the input type is always promoted to `double`.

czgdp1807 commented:
Hi. Thanks a lot for the reviews. I will be addressing them by pushing changes tomorrow morning.

czgdp1807 commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1710,6 +1712,61 @@ def test_mode_all_but_economic(self):
+        assert_(q.shape[-2:] == (m, m))
+        assert_(r.shape[-2:] == (m, n))
+        assert_almost_equal(matmul(q, r), a)
+        assert_almost_equal(swapaxes(q, -1, -2).conj(), np.linalg.inv(q))
```
Yes. In fact, the suggested one is better. I will try it. Thank you.

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3013,105 +3010,77 @@ init_@lapack_func@(GELSD_PARAMS_t *params,
             goto error;
 
-        work_count = (fortran_int)work_size_query.r;
+        work_count_r = (fortran_int) ((@ftyp@*)params->WORK)->r;
+        work_count_i = (fortran_int) ((@ftyp@*)params->WORK)->i;
```
Just wondering, do you have seen a case where the imaginary part of the first `work` item actually contributes to the `work_count`? When `call_@lapack_func@` sets `work(1)`, I cannot see a reason why would it have both non-trivial real and imaginary parts...

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,26 +3159,828 @@ static void
+            goto error;
+
+        work_count_r = (fortran_int) ((@ftyp@*)params->WORK)->r;
+        work_count_i = (fortran_int) ((@ftyp@*)params->WORK)->i;
```
Why would `work_count_i` be non-zero here? See above.

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,26 +3159,828 @@ static void
+            linearize_@TYPE@_matrix(params.TAU, args[1], &tau_in);
+            not_ok = call_@lapack_func@(&params);
+            if (!not_ok) {
+                delinearize_@TYPE@_matrix(args[1], params.TAU, &tau_in);
```
Is delinearize of `tau_in` necessary? `tau_in` is input only buffer.

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,26 +3159,828 @@ static void
+
+        BEGIN_OUTER_LOOP_3
+            int not_ok;
+            linearize_@TYPE@_matrix(params.A, args[0], &a_in);
+            memcpy(params.Q, params.A, safe_n * safe_m * sizeof(@ftyp@));
```
Could this be replaced by
```
linearize_@TYPE@_matrix(params.Q, args[0], &a_in);
```
to avoid memcpy?

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,26 +3159,828 @@ static void
+                delinearize_@TYPE@_matrix(args[2], params.Q, &q_out);
+            } else {
+                error_occurred = 1;
+                nan_@TYPE@_matrix(args[1], &tau_in);
```
`tau_in` is input, why change it?

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,26 +3159,828 @@ static void
+}
+
+static void
+@TYPE@_lstsq(char **args, npy_intp const *dimensions, npy_intp const *steps,
```
Is exposing `@TYPE@_lstsq` required for this PR that ought to be about vectorized QR?
If not, next time implement only required functionality per PR.
Or this about github failing to produce the correct diff?

czgdp1807 replied:
> is this about github failing to produce the correct diff?

I think this is the case because I didn't change anything existing, just added code for `qr`.

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3235,26 +3159,828 @@ static void
+                delinearize_@TYPE@_matrix(args[2], params.Q, &q_out);
+            } else {
+                error_occurred = 1;
+                nan_@TYPE@_matrix(args[1], &tau_in);
```
`tau_in` is input, does it require nan-fill?

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -3125,105 +3094,60 @@ init_@lapack_func@(GELSD_PARAMS_t *params,
-                nan_@REALTYPE@_matrix(args[4], &r_out);
-                *(npy_int*) args[5] = -1;
-                nan_@REALTYPE@_matrix(args[6], &s_out);
+                nan_@TYPE@_matrix(args[0], &a_in);
```
Does this mean that in the case of error, the user will lose its input data in `a`? If so, suggest not doing that.

czgdp1807 replied:
In the rest of the file too, input is not nan-filled. I have removed this line. Thanks for pointing out.

pearu commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -2834,170 +2868,133 @@ static void
 
     work = mem_buff2;
-    iwork = work + work_size;
+    memset(work, 0, work_size);
```
Is this initialization of `work` space needed? If not, suggest not initializing such scratch spaces.

pearu reviewed (changes requested):
This review part covers the C code. My comments are mostly optimization nits and clarifications.

PS: github did not do a good job in displaying diffs this time..

czgdp1807 committed:
Addressed testing reviews

czgdp1807 committed:
addressed C reviews

czgdp1807 committed:
fixed linting issues

czgdp1807 commented:
I have addressed the reviews and the tests seem to pass. Please let me know if there any other concerns or there is something more that can be done. Thank you.

czgdp1807 committed:
removed redundant functions

pearu reviewed (changes requested):
I have a couple of optimization nits and a BC question.

In addition, it is not clear to me why to define function pairs, qr_r_raw_m and qr_r_raw_n, when they both use the same C implementation. The only difference appears in gufunc descriptor signatures: `"(m,n),(m)->(m,m)"` vs `"(m,n),(n)->(m,n)"`.
Would it be possible join these signatures to "(m, n),(k)->(m, k)" and let the implementation to check that `k == min(m, n)`?
The same note applies to other function pairs with `_m` and `_n` sufficies.

czgdp1807 commented:
> In addition, it is not clear to me why to define function pairs, qr_r_raw_m and qr_r_raw_n, when they both use the same C implementation. The only difference appears in gufunc descriptor signatures: `"(m,n),(m)->(m,m)"` vs `"(m,n),(n)->(m,n)"`.
> Would it be possible join these signatures to "(m, n),(k)->(m, k)" and let the implementation to check that `k == min(m, n)`?
> The same note applies to other function pairs with `_m` and `_n` sufficies.

Here's the thread - https://github.com/numpy/numpy/pull/19151#discussion_r642999021 - which covers the same.

pearu commented:
> Here's the thread - [#19151 (comment)](https://github.com/numpy/numpy/pull/19151#discussion_r642999021) - which covers the same.

OK, we have:
- `qr_r_raw_m` and `qr_r_raw_n` with signatures `"(m,n)->(m)" and "(m,n)->(n)"`, respectively
- `qr_reduced_m` and `qr_reduced_n` with signatures `"(m,n),(m)->(m,m)"` and `"(m,n),(n)->(m,n)"`, respectively
- `qr_complete_m` and `qr_complete_n` with signatures `"(m,n),(m)->(m,m)"` and `"(m,n),(n)->(m,m)"`, respectively.

So:
- nothing can be done regarding `qr_r_raw_*`
- `qr_reduced_*` signatures can be merged into `"(m, n), (k) -> (m, k)"`, notice that the comment in the referenced thread does not apply here: all rhs dimensions are present in the lhs
- `qr_complete_*` signatures can be merged into `"(m, n), (k) -> (m, m)"`, again the thread comment does not apply here.

pearu commented on numpy/linalg/linalg.py:
```diff
@@ -904,83 +921,68 @@ def qr(a, mode='reduced'):
+    # same as number of rows, and if the mode is reduced,
+    # then it is the minimum of number of rows and columns.
     if mode == 'complete' and m > n:
         mc = m
-        q = empty((m, m), t)
+        if m <= n:
+            gufunc = _umath_linalg.qr_complete_m
+        else:
+            gufunc = _umath_linalg.qr_complete_n
```
This code looks buggy: the outer `if`-statement has the `m > n` condition, hence only the inner `else` is reachable.
After a fix, could you add a test for complete qr for which `m <= n` holds?

czgdp1807 replied:
It seems like tests already cover all the three cases (m < n, m > n and m = n for all the modes), https://github.com/numpy/numpy/pull/19151/files#diff-b8e9b23ac61b116cd8c650afceac37700ef9f6140cb654a110a22d0d1564df53R1761
The case might be that we never need, `qr_complete_m`. We can remove the `if-else` al-together and just call `qr_complete_n`.

pearu replied:
Sounds good. So, `qr_complete_m` can be removed and `qr_complete_n` renamed to `qr_complete` while having a check that ensures `m > n`.

pearu reviewed (changes requested):
Bug report.

czgdp1807 commented on numpy/linalg/linalg.py:
```diff
@@ -904,83 +921,68 @@ def qr(a, mode='reduced'):
 
     if mode == 'raw':
-        return a, tau
+        q = transpose(a)
+        if t != result_t:
+            q = q.astype(result_t, copy=False)
+            tau = tau.astype(result_t, copy=False)
+        return wrap(q), tau
 
     if mode == 'economic':
         if t != result_t :
             a = a.astype(result_t, copy=False)
-        return wrap(a.T)
+        return wrap(a)
```
AFAICT, eariler the results from Fortran routines were returned to Python, so CMO in Fortran was converted to RMO via the transpose.
However, now, results from Fortran routines are sent to C and then to Python. So, CMO to RMO conversions happens in C and hence, there isn't a requirement to do the same in Python.
In simple words, I'd say, if we were not taking the transpose in Python then take it, and if we were then do not take it now.

pearu replied:
Thanks for the explanation, it makes sense. I also confirm this via local testing.

czgdp1807 committed:
removed redudancy in code

czgdp1807 commented:
I have addressed all the reviews. Let me know if there's something else to be done. Thanks.

pearu reviewed (approved):
LGTM, thanks @czgdp1807 !

czgdp1807 commented:
Thanks @pearu for the feedbacks and thorough reviews.

mattip commented:
Thanks @czgdp1807
